// nestjs-keycloak-admin/dist/main/guards/role.guard.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var RoleGuard_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoleGuard = void 0;
const common_1 = require("@nestjs/common");
const service_1 = require("../service");
const core_1 = require("@nestjs/core");
const public_decorator_1 = require("../decorators/public.decorator");
const roles_decorator_1 = require("../decorators/roles.decorator");
const constants_1 = require("../constants");
let RoleGuard = RoleGuard_1 = class RoleGuard {
    constructor(keycloak, reflector) {
        this.keycloak = keycloak;
        this.reflector = reflector;
        this.logger = new common_1.Logger(RoleGuard_1.name);
    }
    getRequest(context) {
        return context.switchToHttp().getRequest();
    }
    async canActivate(context) {
        var _a;
        const isPublic = this.reflector.get(public_decorator_1.META_PUBLIC, context.getHandler());
        // Emit process, if endpoint is public.
        if (isPublic) {
            return true;
        }
        const request = this.getRequest(context);
        const roleDefinition = this.reflector.get(roles_decorator_1.META_ROLES, context.getHandler());
        // If no resource type is defined as class decorator, emit.
        if (!((_a = roleDefinition === null || roleDefinition === void 0 ? void 0 : roleDefinition.roles) === null || _a === void 0 ? void 0 : _a.length)) {
            return true;
        }
        // If no access token is defined, probably auth guard failed to load.
        if (!request.accessToken) {
            throw new common_1.UnauthorizedException();
        }
        try {
            const grant = await this.keycloak.connect.grantManager.createGrant({
                access_token: request.accessToken,
            });
            const granted = this.isAccessGranted(roleDefinition, grant.access_token);
            granted
                ? this.logger.verbose(`Access granted for role(s).`)
                : this.logger.verbose(`Access denied due to missing role(s).`);
            return granted;
        }
        catch (error) {
            this.logger.error(`Error occured validating roles`, error);
        }
        throw new common_1.UnauthorizedException();
    }
    isAccessGranted(roleDefinition, accessToken) {
        if (roleDefinition.mode == constants_1.RoleMatchingMode.ANY) {
            return roleDefinition.roles.some((role) => accessToken === null || accessToken === void 0 ? void 0 : accessToken.hasRole(role));
        }
        return roleDefinition.roles.every((role) => accessToken === null || accessToken === void 0 ? void 0 : accessToken.hasRole(role));
    }
};
RoleGuard = RoleGuard_1 = __decorate([
    (0, common_1.Injectable)(),
    __param(0, (0, common_1.Inject)(service_1.KeycloakService)),
    __metadata("design:paramtypes", [service_1.KeycloakService,
        core_1.Reflector])
], RoleGuard);
exports.RoleGuard = RoleGuard;
