// nestjs-keycloak-admin/dist/module/guards/auth.guard.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var AuthGuard_1;
import { Injectable, UnauthorizedException, Logger, Inject, } from '@nestjs/common';
import { KeycloakService } from '../service';
import { Reflector } from '@nestjs/core';
import { META_PUBLIC } from '../decorators/public.decorator';
let AuthGuard = AuthGuard_1 = class AuthGuard {
    keycloak;
    reflector;
    logger = new Logger(AuthGuard_1.name);
    constructor(keycloak, reflector) {
        this.keycloak = keycloak;
        this.reflector = reflector;
    }
    getRequest(context) {
        return context.switchToHttp().getRequest();
    }
    async canActivate(context) {
        const isPublic = this.reflector.get(META_PUBLIC, context.getHandler());
        if (isPublic) {
            return true;
        }
        const request = this.getRequest(context);
        const jwt = this.extractJwt(request.headers);
        try {
            const result = await this.keycloak.connect.grantManager.validateAccessToken(jwt);
            if (typeof result === 'string') {
                request.user = await this.keycloak.connect.grantManager.userInfo(jwt);
                request.accessToken = jwt;
                return true;
            }
        }
        catch (error) {
            this.logger.warn(`Error occurred validating token`, error);
        }
        throw new UnauthorizedException();
    }
    extractJwt({ authorization }) {
        if (!authorization) {
            throw new UnauthorizedException();
        }
        const [type, payload] = authorization.split(' ');
        if (type.toLowerCase() !== 'bearer') {
            throw new UnauthorizedException();
        }
        return payload;
    }
};
AuthGuard = AuthGuard_1 = __decorate([
    Injectable(),
    __param(0, Inject(KeycloakService)),
    __metadata("design:paramtypes", [KeycloakService,
        Reflector])
], AuthGuard);
export { AuthGuard };
