// nestjs-keycloak-admin/dist/module/service.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var KeycloakService_1;
import { Logger, Global } from '@nestjs/common';
import AdminClient from '@keycloak/keycloak-admin-client';
import { Issuer } from 'openid-client';
import { resolve } from 'url';
import { ResourceManager } from './lib/resource-manager';
import { PermissionManager } from './lib/permission-manager';
import KeycloakConnect from 'keycloak-connect';
import { RequestManager } from './lib/request-manager';
let KeycloakService = KeycloakService_1 = class KeycloakService {
    logger = new Logger(KeycloakService_1.name);
    tokenSet;
    issuerClient;
    baseUrl;
    requestManager;
    umaConfiguration;
    options;
    connect;
    permissionManager;
    resourceManager;
    client;
    constructor(options) {
        if (!options.baseUrl.startsWith('http')) {
            throw new Error(`Invalid base url. It should start with either http or https.`);
        }
        this.options = options;
        this.baseUrl = resolve(options.baseUrl, `/auth/realms/${options.realmName}`);
        const keycloak = new KeycloakConnect({}, {
            resource: this.options.clientId,
            realm: this.options.realmName,
            'auth-server-url': resolve(this.options.baseUrl, '/auth'),
            secret: this.options.clientSecret,
        });
        keycloak.accessDenied = (req, _res, next) => {
            req.accessDenied = true;
            next();
        };
        this.connect = keycloak;
        this.client = new AdminClient({
            baseUrl: this.options.baseUrl,
            realmName: this.options.realmName,
        });
        this.requestManager = new RequestManager(this, this.baseUrl);
    }
    async initialize() {
        if (this.umaConfiguration) {
            return;
        }
        const { clientId, clientSecret } = this.options;
        const { data } = await this.requestManager.get('/.well-known/uma2-configuration');
        this.umaConfiguration = data;
        this.resourceManager = new ResourceManager(this, data.resource_registration_endpoint);
        this.permissionManager = new PermissionManager(this, data.token_endpoint);
        const keycloakIssuer = await Issuer.discover(data.issuer);
        this.issuerClient = new keycloakIssuer.Client({
            client_id: clientId,
            client_secret: clientSecret,
        });
        this.tokenSet = await this.issuerClient.grant({
            clientId,
            clientSecret,
            grant_type: 'client_credentials',
        });
        if (this.tokenSet?.access_token)
            this.client.setAccessToken(this.tokenSet?.access_token);
        if (this.tokenSet.expires_at) {
            this.logger.verbose(`Initial token expires at ${this.tokenSet.expires_at}`);
        }
    }
    async refreshGrant() {
        if (this.tokenSet && !this.tokenSet.expired()) {
            return this.tokenSet;
        }
        if (!this.tokenSet) {
            this.logger.warn(`Missing token set on refreshGrant.`);
            return null;
        }
        const { refresh_token } = this.tokenSet;
        if (!refresh_token) {
            this.logger.debug(`Refresh token is missing. Reauthenticating.`);
            this.tokenSet = await this.issuerClient?.grant({
                clientId: this.options.clientId,
                clientSecret: this.options.clientSecret,
                grant_type: 'client_credentials',
            });
            if (this.tokenSet?.access_token)
                this.client.setAccessToken(this.tokenSet?.access_token);
            return this.tokenSet;
        }
        this.logger.debug(`Refreshing grant token`);
        this.tokenSet = await this.issuerClient?.refresh(refresh_token);
        return this.tokenSet;
    }
};
KeycloakService = KeycloakService_1 = __decorate([
    Global(),
    __metadata("design:paramtypes", [Object])
], KeycloakService);
export { KeycloakService };
